GTG/gtk/preferences.py

```diff
@@ -17,304 +17,54 @@
+    STACK_CHILD_TITLE = {
+            'general': 'General Preferences'
+        }
 
     def __init__(self, req, vmanager):
         self.req = req
         self.config = self.req.get_config('browser')
         builder = Gtk.Builder()
-        builder.add_from_file(ViewConfig.PREFERENCES_UI_FILE)
+        builder.add_from_file(self.PREFERENCES_UI_FILE)
 
-        self.dialog = builder.get_object("PreferencesDialog")
-        self.dialog.set_title(_("Preferences - %s" % info.NAME))
-        self.pref_autostart = builder.get_object("pref_autostart")
-        self.pref_show_preview = builder.get_object("pref_show_preview")
-        self.bg_color_enable = builder.get_object("bg_color_enable")
-        self.hbox1 = builder.get_object("hbox1")
-        self.shortcut_button = builder.get_object("shortcut_button")
+        self.window = builder.get_object("Preferences")
 
-        self.shortcut = ShortcutWidget(builder)
-        help.add_help_shortcut(self.dialog, "preferences")
+        builder.connect_signals(self)
 
-        self.fontbutton = builder.get_object("fontbutton")
-        self.timer = vmanager.timer
-        self.color_invalid = Color(50000, 0, 0)
-        self.refresh_hour = builder.get_object("hour")
-        self.refresh_mins = builder.get_object("min")
-        editor_font = self.config.get("font_name")
-        if editor_font == "":
-            font = self.dialog.get_style_context().get_font(
-                Gtk.StateFlags.NORMAL)
-            editor_font = font.to_string()
-        self.fontbutton.set_font_name(editor_font)
+        self.headerbar = builder.get_object("right_header_bar")
+        self.stack = builder.get_object("stack")
 
-        builder.connect_signals({
-                                'on_pref_autostart_toggled':
-                                self.on_autostart_toggled,
-                                'on_pref_show_preview_toggled':
-                                self.toggle_preview,
-                                'on_bg_color_toggled':
-                                self.on_bg_color_toggled,
-                                'on_prefs_help':
-                                self.on_help,
-                                'on_prefs_close':
-                                self.on_close,
-                                'on_PreferencesDialog_delete_event':
-                                self.on_close,
-                                'on_fontbutton_font_set':
-                                self.on_font_change,
-                                'on_shortcut_button_toggled':
-                                self.shortcut.on_shortcut_toggled,
-                                'on_valid_check':
-                                self.valid_check,
-                                })
+        self.general_pref = GeneralPreferences(req, vmanager)
+        self.stack.add_titled(self.general_pref.get_ui(),
+                              "general", _("General Preferences"))
```

Comment:
OK, this doesn't work. You have the string "General Preferences" twice in this file. If we decide to change it, they might get out of sync. What about to defining those attributes on the class instead? Something along lines (pseudocode):

```python
class GeneralPreferences:
   name = "general"
   title = _("General Preferences")

# ...
class Preferences:
  def __init__(self):
     # ...
     self.pages = {}
     self.add_page(GeneralPreferences(req, vmanager))

  def add_page(self, page):
     self.pages[page.name] = page
     self.stack.add_titled(page.get_ui(), page.name, page.title)

  def on_sidebar_change(self, widget, data=None):
    page = self.pages[self.stack.get_visible_child_name()]
    self.headerbar.set_title(page.title)
```

It would mean that you can easily add plugins and synchronization:

```python
class PluginsPreferences:
   name = "plugins"
   title = _("Plugins")

class SynchronizationPreferences:
   name = "synchronization"
   title = _("Synchronization")
```